tracer/src/Datadog.Trace/Telemetry/Collectors/IntegrationTelemetryCollector.cs

```diff
@@ -0,0 +1,163 @@
+                {
+                    _integrationsById[i].WasExplicitlyDisabled = 1;
+                }
+            }
+
+            SetHasChanges();
+        }
+
+        /// <summary>
+        /// Should be called when an integration is first executed (not necessarily successfully)
+        /// </summary>
+        public void IntegrationRunning(IntegrationId integrationId)
+        {
+            ref var value = ref _integrationsById[(int)integrationId];
+            if (value.WasExecuted == 1)
+            {
+                return;
+            }
+
+            var previousValue = Interlocked.Exchange(ref value.WasExecuted, 1);
+            if (previousValue == 0)
+            {
+                SetHasChanges();
+            }
+        }
+
+        /// <summary>
+        /// Should be called when an integration successfully generates a span
+        /// </summary>
+        public void IntegrationGeneratedSpan(IntegrationId integrationId)
+        {
+            ref var value = ref _integrationsById[(int)integrationId];
+            if (value.WasExecuted == 1 && value.HasGeneratedSpan == 1)
+            {
+                return;
+            }
+
+            var previousWasExecuted = Interlocked.Exchange(ref value.WasExecuted, 1);
+            var previousHasGeneratedSpan = Interlocked.Exchange(ref value.HasGeneratedSpan, 1);
+
+            if (previousWasExecuted == 0 || previousHasGeneratedSpan == 0)
+            {
+                SetHasChanges();
+            }
+        }
+
+        public void IntegrationDisabledDueToError(IntegrationId integrationId, string error)
+        {
+            ref var value = ref _integrationsById[(int)integrationId];
+            if (value.Error is not null)
+            {
+                return;
+            }
+
+            var previousValue = Interlocked.Exchange(ref value.Error, error);
+            if (previousValue is null)
+            {
+                SetHasChanges();
+            }
+        }
+
+        public bool HasChanges()
+        {
+            return _hasChangesFlag == 1;
+        }
+
+        /// <summary>
+        /// Get the latest data to send to the intake.
+        /// </summary>
+        /// <returns>Null if there are no changes, or the collector is not yet initialized</returns>
+        public ICollection<IntegrationTelemetryData> GetData()
+        {
+            var hasChanges = Interlocked.CompareExchange(ref _hasChangesFlag, 0, 1) == 1;
+            if (!hasChanges)
+            {
+                return null;
+            }
+
+            return _integrationsById
+                  .Select(
+                       integration => new IntegrationTelemetryData
+                       {
+                           Name = integration.Name,
+                           Enabled = integration.HasGeneratedSpan > 0
+                                  && integration.WasExplicitlyDisabled == 0,
+                           AutoEnabled = integration.WasExecuted > 0,
```

Comment:
Using the result of `integration.WasExecuted` for the `auto_enabled` field doesn't seem entirely correct to me. I think it would be more accurate to have a prepopulated array that has a 1:1 mapping of integrationId <=> AutoEnabled

Reply:
While potentially technically more correct, that doesn't give us any extra information does it? We will always know which `IntegrationId`s _could_ be used, this way gives us some extra information?

Reply:
Yeah you're right, we already know what integrations are on by default based on the Tracer version so my suggestion wouldn't give us any new insights